### ISSUE TITLE
Suggest fs::create_dir_all over fs::create_dir

### ISSUE THREAD

**Issue by Jake-Shadle** [rust-lang/rust-clippy#5950]
### What it does

`std::fs::create_dir_all` is a superset of `std::fs::create_dir`, and should be suggested over it, as in (most?) scenarios, `create_dir_all` is what the programmer actually wants.

### Categories (optional)

- Kind: pedantic? maybe? not sure really, it's also kind of correctness to some degree

For example:
- `create_dir` is really only useful as a way to generate errors when 1 or more parent directories don't exist
- Much more reliable in the face of changing directory structures.

### Drawbacks

`create_dir_all` needs to walk up the directory chain and back down to create all parent directories, but it's no less efficient than what the programmer would have to do themselves if they manually implemented create_dir_all.

Uhh, it's also longer to type. :wink: 

### Example

```rust
std::fs::create_dir(&dir_name)?;
```

Could be written as:

```rust
std::fs::create_dir_all(&dir_name)?;
```

**Comment by flip1995** [comment]
This lint can only be implemented as a `restriction` lint, since we can't possible detect if `create_dir_all` was meant instead of `create_dir`.

**Comment by matsujika** [comment]
Can I give it a try?

**Comment by Jake-Shadle** [comment]
Thanks @matsujika!

**Comment by nazar-pc** [comment]
More than once I've seen people having issues with disk mounts and `create_dir_all()` resulting in storing files on their system drive by accident, filling it up and crashing the whole OS as the result.

I found this rule to be more harmful than useful.

I'd say the opposite is most often desirable: using `create_dir()` rather than dangerous `create_dir_all()`. Though I don't think using `create_dir_all()` is wrong in certain cases either.

Very questionable lint IMHO.
